src/audio_device.c

```diff
@@ -46,125 +42,174 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <assert.h>
-
-#define inline__ inline __attribute__((always_inline))
+#include <math.h>
 
 extern struct user_settings *user_settings;
+extern struct Winthread Winthread;
 
+typedef struct FrameInfo {
+    uint32_t samples_per_frame;
+    uint32_t sample_rate;
+    bool stereo;
+} FrameInfo;
+
+/* A virtual input/output device, abstracting the currently selected openal
+ * device (which may change during the lifetime of the virtual device).
+ * We refer to a virtual device as a "device", and refer to an underlying
+ * openal device as an "al_device".
+ * Multiple virtual devices may be open at once; the callback of each virtual
+ * input device has data captured from the input al_device passed to it, and
+ * each virtual output device acts as a source for the output al_device.
+ */
 typedef struct Device {
-    ALCdevice  *dhndl;                     /* Handle of device selected/opened */
-    ALCcontext *ctx;                       /* Device context */
-    DataHandleCallback cb;                 /* Use this to handle data from input device usually */
-    void *cb_data;                         /* Data to be passed to callback */
-    int32_t friend_number;                      /* ToxAV friend number */
-
-    uint32_t source, buffers[OPENAL_BUFS]; /* Playback source/buffers */
-    uint32_t ref_count;
-    int32_t selection;
-    bool enable_VAD;
+    bool active;
     bool muted;
-    pthread_mutex_t mutex[1];
-    uint32_t sample_rate;
-    uint32_t frame_duration;
-    int32_t sound_mode;
-#ifdef AUDIO
-    float VAD_treshold;                    /* 40 is usually recommended value */
-#endif
+
+    FrameInfo frame_info;
+
+    // used only by input devices:
+    DataHandleCallback cb;
+    void *cb_data;
+    float VAD_threshold;
+    uint32_t VAD_samples_remaining;
+
+    // used only by output devices:
+    uint32_t source, buffers[OPENAL_BUFS];
```

Comment:
I'm not a fan of multiple declarations on one line, but even less of a fan if the declarators declare names of different types, in this case `uint32_t` and `uint32_t[OPENAL_BUFS]`.